presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java

```diff
@@ -0,0 +1,235 @@
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class ElasticsearchClient
+{
+    private static final Logger log = Logger.get(ElasticsearchClient.class);
+
+    private final Map<SchemaTableName, ElasticsearchTableDescription> tableDescriptions;
+    private final Map<String, Client> clients = new HashMap<String, Client>();
+    private final Duration timeout;
+
+    @Inject
+    public ElasticsearchClient(Map<SchemaTableName, ElasticsearchTableDescription> descriptions, Duration requestTimeout)
+            throws IOException
+    {
+        tableDescriptions = requireNonNull(descriptions, "Elasticsearch Table Description is null");
+        timeout = requireNonNull(requestTimeout, "Elasticsearch request timeout is null");
+
+        for (Map.Entry<SchemaTableName, ElasticsearchTableDescription> entry : tableDescriptions.entrySet()) {
+            ElasticsearchTableDescription tableDescription = entry.getValue();
+            if (!clients.containsKey(tableDescription.getClusterName())) {
+                Settings settings = Settings.builder().put("cluster.name", tableDescription.getClusterName()).build();
+                TransportClient client = new PreBuiltTransportClient(settings).addTransportAddress(new InetSocketTransportAddress(InetAddress.getByName(tableDescription.getHostAddress()), tableDescription.getPort()));
+                clients.put(tableDescription.getClusterName(), client);
+            }
+        }
+    }
+
+    public List<String> listSchemas()
+    {
+        return tableDescriptions.keySet().stream().map(SchemaTableName::getSchemaName).collect(toImmutableList());
+    }
+
+    public List<SchemaTableName> listTables(String schemaNameOrNull)
+    {
+        return tableDescriptions.keySet().stream().filter(schemaTableName -> schemaNameOrNull == null || schemaTableName.getSchemaName().equals(schemaNameOrNull)).collect(toImmutableList());
+    }
+
+    public ElasticsearchTableDescription getTable(String schemaName, String tableName)
+    {
+        requireNonNull(schemaName, "schemaName is null");
+        requireNonNull(tableName, "tableName is null");
+        for (Map.Entry<SchemaTableName, ElasticsearchTableDescription> entry : tableDescriptions.entrySet()) {
+            buildColumns(entry.getValue());
+        }
+        return tableDescriptions.get(new SchemaTableName(schemaName, tableName));
+    }
+
+    public ClusterSearchShardsResponse getSearchShards(ElasticsearchTableDescription tableDescription)
+    {
+        Client client = clients.get(tableDescription.getClusterName());
+        verify(client != null);
+        ClusterSearchShardsRequest request = new ClusterSearchShardsRequest(tableDescription.getIndex());
+        return client.admin().cluster().searchShards(request).actionGet(timeout.toMillis());
+    }
+
+    private void buildColumns(ElasticsearchTableDescription tableDescription)
```

Comment:
this method is pretty complicated, we can extract another method to populate the columns from the mappings: https://gist.github.com/nezihyigitbasi/52fbd8056a80de2bd17b50a49523a5db